File: Python/Problems/Problems/problem2452.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set
from collections import Counter

from .problem import Problem


class Problem2452(Problem):
    def test(self) -> bool:
        queries = ["word", "note", "ants", "wood"]
        dictionary = ["wood", "joke", "moat"]

        expected = ["word", "note", "wood"]

        result = self.twoEditWords(queries, dictionary)

        print1DArray(result)

        return expected == result

    def twoEditWords(self, queries: List[str], dictionary: List[str]) -> List[str]:
        n = dict()
        result = []
        for query in queries:
            found = False
            for word in dictionary:
                if (len(query) != len(word)):
                    continue
                l = len(query)
                diff = 0
                i = 0
                while i < l and diff <= 2:
                    if word[i] != query[i]:
                        diff += 1
                    i += 1
                if diff <= 2:
                    found = True
                    break
            if found:
                result.append(query)
        return result

File: C++/Problems/Problems/problem_2452.cpp
#include "problem_2452.h"

bool problem_2452::test() {
    std::vector<std::string> queries { "word", "note", "ants", "wood" };
    std::vector<std::string> dictionary { "wood", "joke", "moat" };

    std::vector<std::string> expected { "word", "note", "wood" };

    auto result = twoEditWords(queries, dictionary);

    print1DVector(result);

    return expected == result;
}

std::vector<std::string> problem_2452::twoEditWords(std::vector<std::string> &queries,
    std::vector<std::string> &dictionary) {
    int n = dictionary.size();
    std::vector<std::string> result;
    for (auto query : queries)
    {
        bool found = false;
        for (auto word : dictionary)
        {
            if (query.size() != word.size())
                continue;
            int l = query.size();
            int diff = 0;
            for (int i = 0; i < l && diff <= 2; ++i)
            {
                if (word[i] != query[i])
                    diff++;
            }
            if (diff <= 2)
            {
                found = true;
                break;
            }
        }
        if (found)
            result.push_back(query);
    }
    return result;
}
